tests/Collection/CollectionFunctionalTest.php

```diff
@@ -170,6 +173,88 @@ function (array $event) {
         );
     }
 
+    /**
+     * @dataProvider provideTypeMapOptionsAndExpectedDocuments
+     */
+    public function testDistinctWithTypeMap(array $typeMap, array $expectedDocuments)
+    {
+        $bulkWrite = new BulkWrite(['ordered' => true]);
+        $bulkWrite->insert([
+            'x' => (object) ['foo' => 'bar'],
+        ]);
+        $bulkWrite->insert(['x' => 4]);
+        $bulkWrite->insert([
+            'x' => (object) ['foo' => ['foo' => 'bar']],
+        ]);
+        $this->manager->executeBulkWrite($this->getNamespace(), $bulkWrite);
+
+        $values = $this->collection->withOptions(['typeMap' => $typeMap])->distinct('x');
+
+        /* This sort callable sorts all scalars to the front of the list. All
+         * non-scalar values are sorted by running json_encode on them and
+         * comparing their string representations.
+         */
+        $sort = function ($a, $b) {
```

Comment:
Would it be easier to just assert that the count of `$expectedDocuments` and `$values` is equal and then loop over `$expectedDocuments` and assert that each of its elements exists in `$values`?

I realize the only scalar is `4` at the moment, but this seems like it could be fragile if additional scalars were added to the test data.

Reply:
I shamelessly copied this from DistinctFunctionalTest, so I had to double-check what it does. This method generally sorts scalars to the top and non-scalars to the bottom, but each group is sorted within itself, so adding multiple scalars shouldn't pose an issue. Non-scalars are encoded to JSON before sorting, so they are sorted by their JSON representation.